src/librustc/mir/mod.rs

```diff
@@ -104,6 +104,15 @@ pub struct Mir<'tcx> {
     /// Return type of the function.
     pub return_ty: Ty<'tcx>,
 
+    /// Yield type of the function, if it is a generator.
+    pub yield_ty: Option<Ty<'tcx>>,
+
+    /// Generator drop glue
+    pub generator_drop: Option<Box<Mir<'tcx>>>,
+
+    /// The layout of a generator. Produced by the state transformation.
+    pub generator_layout: Option<GeneratorLayout<'tcx>>,
```

Comment:
I'm worried this might create some cycles if MIR optimizations start accessing type layouts, but I guess that's life and we could find a fix.